### PR TITLE
Return json on 500 error from API

### PR DESCRIPTION
**Pull request by 6543**
if ctx.InternalServerError(err) is used it will return the default 500 html page

on API a json Error is expected

### CODE REVIEW

**Comment by 6543** [comment]
just noticed this behaviour - it was not documented at all - of coures if you do a code lookup you can see it too :)

**Comment by 6543** [comment]
@techknowlogick can we backport this somehow?

**Comment by techknowlogick** [comment]
whoops. label didn't apply at first. It's there now:)

File: routers/api/v1/misc/markdown.go [hunk 1]
```diff
@@ -66,20 +66,20 @@ func Markdown(ctx *context.APIContext, form api.MarkdownOption) {
 		if form.Wiki {
 			_, err := ctx.Write([]byte(markdown.RenderWiki(md, urlPrefix, meta)))
 			if err != nil {
-				ctx.InternalServerError(err)
+				ctx.Error(http.StatusInternalServerError, "", err)
```
Comment (6543) [inline routers/api/v1/misc/markdown.go]:
```suggestion
				ctx.Error(http.StatusInternalServerError, "ctx.Write", err)
```

File: routers/api/v1/misc/markdown.go [hunk 2]
```diff
@@ -66,20 +66,20 @@ func Markdown(ctx *context.APIContext, form api.MarkdownOption) {
...
 				return
 			}
 		} else {
 			_, err := ctx.Write(markdown.Render(md, urlPrefix, meta))
 			if err != nil {
-				ctx.InternalServerError(err)
+				ctx.Error(http.StatusInternalServerError, "", err)
```
Comment (6543) [inline routers/api/v1/misc/markdown.go]:
```suggestion
				ctx.Error(http.StatusInternalServerError, "markdown.Render", err)
```

File: routers/api/v1/misc/markdown.go [hunk 3]
```diff
@@ -66,20 +66,20 @@ func Markdown(ctx *context.APIContext, form api.MarkdownOption) {
...
 				return
 			}
 		}
 	default:
 		_, err := ctx.Write(markdown.RenderRaw([]byte(form.Text), "", false))
 		if err != nil {
-			ctx.InternalServerError(err)
+			ctx.Error(http.StatusInternalServerError, "", err)
```
Comment (6543) [inline routers/api/v1/misc/markdown.go]:
```suggestion
			ctx.Error(http.StatusInternalServerError, "markdown.RenderRaw", err)
```

File: routers/api/v1/misc/markdown.go [hunk 4]
```diff
@@ -114,7 +114,7 @@ func MarkdownRaw(ctx *context.APIContext) {
 	}
 	_, err = ctx.Write(markdown.RenderRaw(body, "", false))
 	if err != nil {
-		ctx.InternalServerError(err)
+		ctx.Error(http.StatusInternalServerError, "", err)
```
Comment (6543) [inline routers/api/v1/misc/markdown.go]:
```suggestion
		ctx.Error(http.StatusInternalServerError, "markdown.RenderRaw", err)
```

File: routers/api/v1/notify/repo.go [hunk 5]
```diff
@@ -65,7 +65,7 @@ func ListRepoNotifications(ctx *context.APIContext) {
 
 	before, since, err := utils.GetQueryBeforeSince(ctx)
 	if err != nil {
-		ctx.InternalServerError(err)
+		ctx.Error(http.StatusInternalServerError, "", err)
```
Comment (6543) [inline routers/api/v1/notify/repo.go]:
```suggestion
		ctx.Error(http.StatusInternalServerError, "GetQueryBeforeSince", err)
```

File: routers/api/v1/notify/repo.go [hunk 6]
```diff
@@ -81,12 +81,12 @@ func ListRepoNotifications(ctx *context.APIContext) {
 	}
 	nl, err := models.GetNotifications(opts)
 	if err != nil {
-		ctx.InternalServerError(err)
+		ctx.Error(http.StatusInternalServerError, "", err)
```
Comment (6543) [inline routers/api/v1/notify/repo.go]:
```suggestion
		ctx.Error(http.StatusInternalServerError, "GetNotifications", err)
```

File: routers/api/v1/notify/repo.go [hunk 7]
```diff
@@ -81,12 +81,12 @@ func ListRepoNotifications(ctx *context.APIContext) {
...
 		return
 	}
 	err = nl.LoadAttributes()
 	if err != nil {
-		ctx.InternalServerError(err)
+		ctx.Error(http.StatusInternalServerError, "", err)
```
Comment (6543) [inline routers/api/v1/notify/repo.go]:
```suggestion
		ctx.Error(http.StatusInternalServerError, "nl.LoadAttributes", err)
```

File: routers/api/v1/notify/repo.go [hunk 8]
```diff
@@ -128,7 +128,7 @@ func ReadRepoNotifications(ctx *context.APIContext) {
 	if len(qLastRead) > 0 {
 		tmpLastRead, err := time.Parse(time.RFC3339, qLastRead)
 		if err != nil {
-			ctx.InternalServerError(err)
+			ctx.Error(http.StatusInternalServerError, "", err)
```
Comment (6543) [inline routers/api/v1/notify/repo.go]:
```suggestion
			ctx.Error(http.StatusInternalServerError, "time.Parse", err)
```

File: routers/api/v1/notify/repo.go [hunk 9]
```diff
@@ -143,14 +143,14 @@ func ReadRepoNotifications(ctx *context.APIContext) {
 	}
 	nl, err := models.GetNotifications(opts)
 	if err != nil {
-		ctx.InternalServerError(err)
+		ctx.Error(http.StatusInternalServerError, "", err)
```
Comment (6543) [inline routers/api/v1/notify/repo.go]:
```suggestion
		ctx.Error(http.StatusInternalServerError, "GetNotifications", err)
```

File: routers/api/v1/notify/repo.go [hunk 10]
```diff
@@ -143,14 +143,14 @@ func ReadRepoNotifications(ctx *context.APIContext) {
...
 		return
 	}
 
 	for _, n := range nl {
 		err := models.SetNotificationStatus(n.ID, ctx.User, models.NotificationStatusRead)
 		if err != nil {
-			ctx.InternalServerError(err)
+			ctx.Error(http.StatusInternalServerError, "", err)
```
Comment (6543) [inline routers/api/v1/notify/repo.go]:
```suggestion
			ctx.Error(http.StatusInternalServerError, "SetNotificationStatus", err)
```

File: routers/api/v1/notify/threads.go [hunk 11]
```diff
@@ -40,7 +40,7 @@ func GetThread(ctx *context.APIContext) {
 		return
 	}
 	if err := n.LoadAttributes(); err != nil {
-		ctx.InternalServerError(err)
+		ctx.Error(http.StatusInternalServerError, "", err)
```
Comment (6543) [inline routers/api/v1/notify/threads.go]:
```suggestion
		ctx.Error(http.StatusInternalServerError, "n.LoadAttributes", err)
```

File: routers/api/v1/notify/threads.go [hunk 12]
```diff
@@ -77,7 +77,7 @@ func ReadThread(ctx *context.APIContext) {
 
 	err := models.SetNotificationStatus(n.ID, ctx.User, models.NotificationStatusRead)
 	if err != nil {
-		ctx.InternalServerError(err)
+		ctx.Error(http.StatusInternalServerError, "", err)
```
Comment (6543) [inline routers/api/v1/notify/threads.go]:
```suggestion
		ctx.Error(http.StatusInternalServerError, "SetNotificationStatus", err)
```

File: routers/api/v1/notify/threads.go [hunk 13]
```diff
@@ -89,7 +89,7 @@ func getThread(ctx *context.APIContext) *models.Notification {
 		if models.IsErrNotExist(err) {
 			ctx.Error(http.StatusNotFound, "GetNotificationByID", err)
 		} else {
-			ctx.InternalServerError(err)
+			ctx.Error(http.StatusInternalServerError, "", err)
```
Comment (6543) [inline routers/api/v1/notify/threads.go]:
```suggestion
			ctx.Error(http.StatusInternalServerError, "GetNotificationByID", err)
```

File: routers/api/v1/notify/user.go [hunk 14]
```diff
@@ -55,7 +55,7 @@ func ListNotifications(ctx *context.APIContext) {
 
 	before, since, err := utils.GetQueryBeforeSince(ctx)
 	if err != nil {
-		ctx.InternalServerError(err)
+		ctx.Error(http.StatusInternalServerError, "", err)
```
Comment (6543) [inline routers/api/v1/notify/user.go]:
```suggestion
		ctx.Error(http.StatusInternalServerError, "GetQueryBeforeSince", err)
```

File: routers/api/v1/notify/user.go [hunk 15]
```diff
@@ -70,12 +70,12 @@ func ListNotifications(ctx *context.APIContext) {
 	}
 	nl, err := models.GetNotifications(opts)
 	if err != nil {
-		ctx.InternalServerError(err)
+		ctx.Error(http.StatusInternalServerError, "", err)
```
Comment (6543) [inline routers/api/v1/notify/user.go]:
```suggestion
		ctx.Error(http.StatusInternalServerError, "GetNotifications", err)
```

File: routers/api/v1/notify/user.go [hunk 16]
```diff
@@ -70,12 +70,12 @@ func ListNotifications(ctx *context.APIContext) {
...
 		return
 	}
 	err = nl.LoadAttributes()
 	if err != nil {
-		ctx.InternalServerError(err)
+		ctx.Error(http.StatusInternalServerError, "", err)
```
Comment (6543) [inline routers/api/v1/notify/user.go]:
```suggestion
		ctx.Error(http.StatusInternalServerError, "nl.LoadAttributes", err)
```

File: routers/api/v1/notify/user.go [hunk 17]
```diff
@@ -107,7 +107,7 @@ func ReadNotifications(ctx *context.APIContext) {
 	if len(qLastRead) > 0 {
 		tmpLastRead, err := time.Parse(time.RFC3339, qLastRead)
 		if err != nil {
-			ctx.InternalServerError(err)
+			ctx.Error(http.StatusInternalServerError, "", err)
```
Comment (6543) [inline routers/api/v1/notify/user.go]:
```suggestion
			ctx.Error(http.StatusInternalServerError, "time.Parse", err)
```

File: routers/api/v1/notify/user.go [hunk 18]
```diff
@@ -121,14 +121,14 @@ func ReadNotifications(ctx *context.APIContext) {
 	}
 	nl, err := models.GetNotifications(opts)
 	if err != nil {
-		ctx.InternalServerError(err)
+		ctx.Error(http.StatusInternalServerError, "", err)
```
Comment (6543) [inline routers/api/v1/notify/user.go]:
```suggestion
		ctx.Error(http.StatusInternalServerError, "GetNotifications", err)
```

File: routers/api/v1/notify/user.go [hunk 19]
```diff
@@ -121,14 +121,14 @@ func ReadNotifications(ctx *context.APIContext) {
...
 		return
 	}
 
 	for _, n := range nl {
 		err := models.SetNotificationStatus(n.ID, ctx.User, models.NotificationStatusRead)
 		if err != nil {
-			ctx.InternalServerError(err)
+			ctx.Error(http.StatusInternalServerError, "", err)
```
Comment (6543) [inline routers/api/v1/notify/user.go]:
```suggestion
			ctx.Error(http.StatusInternalServerError, "SetNotificationStatus", err)
```

File: routers/api/v1/org/team.go [hunk 20]
```diff
@@ -213,7 +213,7 @@ func EditTeam(ctx *context.APIContext, form api.EditTeamOption) {
 
 	team := ctx.Org.Team
 	if err := team.GetUnits(); err != nil {
-		ctx.InternalServerError(err)
+		ctx.Error(http.StatusInternalServerError, "", err)
```
Comment (6543) [inline routers/api/v1/org/team.go]:
```suggestion
		ctx.Error(http.StatusInternalServerError, "GetUnits", err)
```

File: routers/api/v1/repo/branch.go [hunk 21]
```diff
@@ -128,7 +128,7 @@ func DeleteBranch(ctx *context.APIContext) {
 
 	isProtected, err := ctx.Repo.Repository.IsProtectedBranch(ctx.Repo.BranchName, ctx.User)
 	if err != nil {
-		ctx.InternalServerError(err)
+		ctx.Error(http.StatusInternalServerError, "", err)
```
Comment (6543) [inline routers/api/v1/repo/branch.go]:
```suggestion
		ctx.Error(http.StatusInternalServerError, "IsProtectedBranch", err)
```

File: routers/api/v1/repo/issue.go [hunk 22]
```diff
@@ -624,11 +624,11 @@ func EditIssue(ctx *context.APIContext, form api.EditIssueOption) {
 	// Refetch from database to assign some automatic values
 	issue, err = models.GetIssueByID(issue.ID)
 	if err != nil {
-		ctx.InternalServerError(err)
+		ctx.Error(http.StatusInternalServerError, "", err)
```
Comment (6543) [inline routers/api/v1/repo/issue.go]:
```suggestion
		ctx.Error(http.StatusInternalServerError, "GetIssueByID", err)
```

File: routers/api/v1/repo/issue.go [hunk 23]
```diff
@@ -624,11 +624,11 @@ func EditIssue(ctx *context.APIContext, form api.EditIssueOption) {
...
 		return
 	}
 	if err = issue.LoadMilestone(); err != nil {
-		ctx.InternalServerError(err)
+		ctx.Error(http.StatusInternalServerError, "", err)
```
Comment (6543) [inline routers/api/v1/repo/issue.go]:
```suggestion
		ctx.Error(http.StatusInternalServerError, "issue.LoadMilestone", err)
```

File: routers/api/v1/repo/issue_comment.go [hunk 24]
```diff
@@ -275,7 +275,7 @@ func GetIssueComment(ctx *context.APIContext) {
 	}
 
 	if err = comment.LoadIssue(); err != nil {
-		ctx.InternalServerError(err)
+		ctx.Error(http.StatusInternalServerError, "", err)
```
Comment (6543) [inline routers/api/v1/repo/issue_comment.go]:
```suggestion
		ctx.Error(http.StatusInternalServerError, "comment.LoadIssue", err)
```

File: routers/api/v1/repo/issue_subscription.go [hunk 25]
```diff
@@ -196,7 +196,7 @@ func CheckIssueSubscription(ctx *context.APIContext) {
 
 	watching, err := models.CheckIssueWatch(ctx.User, issue)
 	if err != nil {
-		ctx.InternalServerError(err)
+		ctx.Error(http.StatusInternalServerError, "", err)
```
Comment (6543) [inline routers/api/v1/repo/issue_subscription.go]:
```suggestion
		ctx.Error(http.StatusInternalServerError, "CheckIssueWatch", err)
```

File: routers/api/v1/repo/issue_tracked_time.go [hunk 26]
```diff
@@ -86,7 +86,7 @@ func ListTrackedTimes(ctx *context.APIContext) {
 	}
 
 	if opts.CreatedBeforeUnix, opts.CreatedAfterUnix, err = utils.GetQueryBeforeSince(ctx); err != nil {
-		ctx.InternalServerError(err)
+		ctx.Error(http.StatusInternalServerError, "", err)
```
Comment (6543) [inline routers/api/v1/repo/issue_tracked_time.go]:
```suggestion
		ctx.Error(http.StatusInternalServerError, "GetQueryBeforeSince", err)
```

File: routers/api/v1/repo/issue_tracked_time.go [hunk 27]
```diff
@@ -491,7 +491,7 @@ func ListTrackedTimesByRepository(ctx *context.APIContext) {
 
 	var err error
 	if opts.CreatedBeforeUnix, opts.CreatedAfterUnix, err = utils.GetQueryBeforeSince(ctx); err != nil {
-		ctx.InternalServerError(err)
+		ctx.Error(http.StatusInternalServerError, "", err)
```
Comment (6543) [inline routers/api/v1/repo/issue_tracked_time.go]:
```suggestion
		ctx.Error(http.StatusInternalServerError, "GetQueryBeforeSince", err)
```

File: routers/api/v1/repo/issue_tracked_time.go [hunk 28]
```diff
@@ -554,7 +554,7 @@ func ListMyTrackedTimes(ctx *context.APIContext) {
...
```
Comment (6543) [inline routers/api/v1/repo/issue_tracked_time.go]:
```suggestion
		ctx.Error(http.StatusInternalServerError, "GetQueryBeforeSince", err)
```

File: routers/api/v1/repo/topic.go [hunk 29]
```diff
@@ -52,7 +52,7 @@ func ListTopics(ctx *context.APIContext) {
 	})
 	if err != nil {
 		log.Error("ListTopics failed: %v", err)
-		ctx.InternalServerError(err)
+		ctx.Error(http.StatusInternalServerError, "", err)
```
Comment (6543) [inline routers/api/v1/repo/topic.go]:
```suggestion
		ctx.Error(http.StatusInternalServerError, "ListTopics", err)
```

File: routers/api/v1/repo/topic.go [hunk 30]
```diff
@@ -115,7 +115,7 @@ func UpdateTopics(ctx *context.APIContext, form api.RepoTopicOptions) {
 	err := models.SaveTopics(ctx.Repo.Repository.ID, validTopics...)
 	if err != nil {
 		log.Error("SaveTopics failed: %v", err)
-		ctx.InternalServerError(err)
+		ctx.Error(http.StatusInternalServerError, "", err)
```
Comment (6543) [inline routers/api/v1/repo/topic.go]:
```suggestion
		ctx.Error(http.StatusInternalServerError, "SaveTopics", err)
```

File: routers/api/v1/repo/topic.go [hunk 31]
```diff
@@ -52,7 +52,7 @@ func ListTopics(ctx *context.APIContext) {
 	})
 	if err != nil {
 		log.Error("ListTopics failed: %v", err)
```
Comment (6543) [inline routers/api/v1/repo/topic.go]:
```suggestion
```

File: routers/api/v1/repo/topic.go [hunk 32]
```diff
@@ -115,7 +115,7 @@ func UpdateTopics(ctx *context.APIContext, form api.RepoTopicOptions) {
 	err := models.SaveTopics(ctx.Repo.Repository.ID, validTopics...)
 	if err != nil {
 		log.Error("SaveTopics failed: %v", err)
```
Comment (6543) [inline routers/api/v1/repo/topic.go]:
```suggestion
```

File: routers/api/v1/repo/topic.go [hunk 33]
```diff
@@ -167,7 +167,7 @@ func AddTopic(ctx *context.APIContext) {
 	})
 	if err != nil {
 		log.Error("AddTopic failed: %v", err)
-		ctx.InternalServerError(err)
+		ctx.Error(http.StatusInternalServerError, "", err)
```
Comment (6543) [inline routers/api/v1/repo/topic.go]:
```suggestion
		ctx.Error(http.StatusInternalServerError, "AddTopic", err)
```

File: routers/api/v1/repo/topic.go [hunk 34]
```diff
@@ -167,7 +167,7 @@ func AddTopic(ctx *context.APIContext) {
 	})
 	if err != nil {
 		log.Error("AddTopic failed: %v", err)
```
Comment (6543) [inline routers/api/v1/repo/topic.go]:
```suggestion
```

File: routers/api/v1/repo/topic.go [hunk 35]
```diff
@@ -180,7 +180,7 @@ func AddTopic(ctx *context.APIContext) {
 	_, err = models.AddTopic(ctx.Repo.Repository.ID, topicName)
 	if err != nil {
 		log.Error("AddTopic failed: %v", err)
```
Comment (6543) [inline routers/api/v1/repo/topic.go]:
```suggestion
```

File: routers/api/v1/repo/topic.go [hunk 36]
```diff
@@ -180,7 +180,7 @@ func AddTopic(ctx *context.APIContext) {
...
-		ctx.InternalServerError(err)
+		ctx.Error(http.StatusInternalServerError, "", err)
```
Comment (6543) [inline routers/api/v1/repo/topic.go]:
```suggestion
		ctx.Error(http.StatusInternalServerError, "AddTopic", err)
```

File: routers/api/v1/repo/topic.go [hunk 37]
```diff
@@ -229,7 +229,7 @@ func DeleteTopic(ctx *context.APIContext) {
 	topic, err := models.DeleteTopic(ctx.Repo.Repository.ID, topicName)
 	if err != nil {
 		log.Error("DeleteTopic failed: %v", err)
-		ctx.InternalServerError(err)
+		ctx.Error(http.StatusInternalServerError, "", err)
```
Comment (6543) [inline routers/api/v1/repo/topic.go]:
```suggestion
		ctx.Error(http.StatusInternalServerError, "DeleteTopic", err)
```

File: routers/api/v1/repo/topic.go [hunk 38]
```diff
@@ -229,7 +229,7 @@ func DeleteTopic(ctx *context.APIContext) {
 	topic, err := models.DeleteTopic(ctx.Repo.Repository.ID, topicName)
 	if err != nil {
 		log.Error("DeleteTopic failed: %v", err)
```
Comment (6543) [inline routers/api/v1/repo/topic.go]:
```suggestion
```

File: routers/api/v1/repo/topic.go [hunk 39]
```diff
@@ -288,7 +288,7 @@ func TopicSearch(ctx *context.APIContext) {
 	})
 	if err != nil {
 		log.Error("SearchTopics failed: %v", err)
```
Comment (6543) [inline routers/api/v1/repo/topic.go]:
```suggestion
```

File: routers/api/v1/repo/topic.go [hunk 40]
```diff
@@ -288,7 +288,7 @@ func TopicSearch(ctx *context.APIContext) {
...
-		ctx.InternalServerError(err)
+		ctx.Error(http.StatusInternalServerError, "", err)
```
Comment (6543) [inline routers/api/v1/repo/topic.go]:
```suggestion
		ctx.Error(http.StatusInternalServerError, "SearchTopics", err)
```

File: routers/api/v1/repo/transfer.go [hunk 41]
```diff
@@ -56,7 +56,7 @@ func Transfer(ctx *context.APIContext, opts api.TransferRepoOption) {
 			ctx.Error(http.StatusNotFound, "GetUserByName", err)
 			return
 		}
-		ctx.InternalServerError(err)
+		ctx.Error(http.StatusInternalServerError, "", err)
```
Comment (6543) [inline routers/api/v1/repo/transfer.go]:
```suggestion
		ctx.Error(http.StatusInternalServerError, "GetUserByName", err)
```

File: routers/api/v1/repo/transfer.go [hunk 42]
```diff
@@ -85,13 +85,13 @@ func Transfer(ctx *context.APIContext, opts api.TransferRepoOption) {
 	}
 
 	if err = repo_service.TransferOwnership(ctx.User, newOwner, ctx.Repo.Repository, teams); err != nil {
-		ctx.InternalServerError(err)
+		ctx.Error(http.StatusInternalServerError, "", err)
```
Comment (6543) [inline routers/api/v1/repo/transfer.go]:
```suggestion
		ctx.Error(http.StatusInternalServerError, "TransferOwnership", err)
```

File: routers/api/v1/repo/transfer.go [hunk 43]
```diff
@@ -85,13 +85,13 @@ func Transfer(ctx *context.APIContext, opts api.TransferRepoOption) {
...
 		return
 	}
 
 	newRepo, err := models.GetRepositoryByName(newOwner.ID, ctx.Repo.Repository.Name)
 	if err != nil {
-		ctx.InternalServerError(err)
+		ctx.Error(http.StatusInternalServerError, "", err)
```
Comment (6543) [inline routers/api/v1/repo/transfer.go]:
```suggestion
		ctx.Error(http.StatusInternalServerError, "GetRepositoryByName", err)
```

File: routers/api/v1/user/app.go [hunk 44]
```diff
@@ -93,7 +93,7 @@ func CreateAccessToken(ctx *context.APIContext, form api.CreateAccessTokenOption
 
 	exist, err := models.AccessTokenByNameExists(t)
 	if err != nil {
-		ctx.InternalServerError(err)
+		ctx.Error(http.StatusInternalServerError, "", err)
```
Comment (6543) [inline routers/api/v1/user/app.go]:
```suggestion
		ctx.Error(http.StatusInternalServerError, "AccessTokenByNameExists", err)
```

**Review by 6543 — COMMENTED** [review]
better error logging

**Comment by silverwind** [comment]
Bit of a shame that this needs to be done in every route. The way I'd handle this in JS routing is to let errors throw in the routes and then process them in a "catch-all" route but I guess with missing exceptions, this is not possible in Golang.

**Review by zeripath — APPROVED** [review]

**Comment by zeripath** [comment]
The other option is for the API context to provide its own internalservererror function

**Comment by 6543** [comment]
we should keep it simple ...

**Comment by zeripath** [comment]
That would be simpler...

in modules/context/api.go add the following function:

```

// InternalServerError responds with an error message to the client with the error as a message and the file and line of the caller.
func (ctx *APIContext) InternalServerError(err error) {
	log.ErrorWithSkip(1, "InternalServerError: %v", err)

	ctx.JSON(http.StatusInternalServerError, APIError{
		Message: err.Error(),
		URL:     setting.API.SwaggerURL,
	})
}
```

And whilst you're at it change the above function to: ```go
// Error responds with an error message to client with given obj as the message.
// If status is 500, also it prints error to log.
func (ctx *APIContext) Error(status int, title string, obj interface{}) {
	var message string
	if err, ok := obj.(error); ok {
		message = err.Error()
	} else {
		message = obj.(string)
	}

	if status == http.StatusInternalServerError {
		log.ErrorWithSkip(1, "%s: %s", title, message)
	}

	ctx.JSON(status, APIError{
		Message: message,
		URL:     setting.API.SwaggerURL,
	})
}
```

File: modules/context/api.go [hunk 45]
```diff
@@ -74,8 +75,8 @@ func (ctx *APIContext) Error(status int, title string, obj interface{}) {
 		message = obj.(string)
```
Comment (zeripath) [inline modules/context/api.go]: Just thinking on this might be nicer

```suggestion
		message = fmt.Sprintf("%s", obj)
```

**Review by zeripath — APPROVED** [review]
Much simpler

**Review by lafriks — CHANGES_REQUESTED** [review]
But won't this reveal internal error messages in api? That would be bad from security point of view

**Comment by silverwind** [comment]
What secrets would they reveal? Does it matter in an open-source application?

**Comment by 6543** [comment]
> But won't this reveal internal error messages in api? That would be bad from security point of view

@lafriks 
they already do it now!!
just in html output and now in json

**Comment by lafriks** [comment]
I don't think actual error message is returned, if it is it should be fixed not to be. Even returning things like paths where Gitea is installed can be dangerous as it could give out information about where files are located. Also knowing what exact unexpected error is happening is so much easier to use it.
This does not matter if app is open source or not.

**Comment by zeripath** [comment]
@lafriks the actual error message is returned currently - but yeah we shouldn't do that.
@6543 add a check like: 

https://github.com/go-gitea/gitea/blob/02a52d683b32e2670f746d898bd0041954390e7b/modules/context/context.go#L180

Before setting the message.

**Review by zeripath — APPROVED** [review]

**Comment by 6543** [comment]
@lafriks I think to hide 500 errors on API will need its own pull to correct 500 errors witch should have other http status codes ...

**Comment by zeripath** [comment]
@6543 to fix the tests I guess if you just ensure that macaron.Env is set to development for them?

**Comment by 6543** [comment]
I would consider this as hotfix only

**Review by lafriks — APPROVED** [review]

**Comment by 6543** [comment]
Close this untill consensus reached, if it is done this way it is kind/breaking

I'll chat about tia on discord and reopen it avterwards again

**Comment by zeripath** [comment]
Hmm... It's not breaking to make security changes - no one should be relying on reading the message of an internal security error.

500s mean that there is a bug.

They should not ever be expected.

**Comment by silverwind** [comment]
I guess you can never make sure 500's don't happen but if they do, they should at least be in expected format (e.g. content-type json).

**Comment by 6543** [comment]
The problem Is: If tests fail they relay on the "message" part of an 500 error. witch mean there exist APIs witch return 500 for **expected** errors witch should be not the case in first place but they are

**Comment by zeripath** [comment]
OK well technically it's a security bug to reveal errors in this fashion as it can be used to probe the server for vulnerabilities - so we have to fix this and can't just ignore it.

**Comment by 6543** [comment]
@zeripath @lafriks I think we should **revert** https://github.com/go-gitea/gitea/pull/11574/commits/8467b2cee674ad205b452780ca88abb1b27643c8 for **this** pull and fix this in a other one avter API endpoints dont return expected 500, else we cant realy backport this

**Comment by zeripath** [comment]
How about we just put the err protection on the InternalServerError route? That way we don't make the situation any worse than we currently have and we can fix the Error(500, ...)s later?

**Comment by 6543** [comment]
@zeripath https://github.com/go-gitea/gitea/pull/11641#issuecomment-634556686 and 497f22689e1244dd6e2d294ace0ad08a09fa4e12

**Review by zeripath — APPROVED** [review]
Shame about the api.Error endpoint - but if you don't have time to fix those right now - at least we can get a fix in for the ostensible reason for this PR.

These changes now at least don't make the situation worse.

**Comment by 6543** [comment]
should we merge 🚀 ?

**Review by guillep2k — APPROVED** [review]

**Review by lafriks — APPROVED** [review]

**Review by lafriks — CHANGES_REQUESTED** [review]
I don't see where InternalServerError function is used anymore :)

**Comment by 6543** [comment]
?

**Comment by zeripath** [comment]
This PR provides a `modules/context.APIContext.InternalServerError(...)` method override for a method in `modules/context.Context`.

An example of a piece of code affected is in routers/api/v1/misc/markdown.go:79:`ctx.InternalServerError(err)`

Where `ctx` is actually a `modules/context.APIContext` and `modules/context.APIContext` is:

```go
// APIContext is a specific macaron context for API service
type APIContext struct {
	*Context
	Org *APIOrganization
}
```

As there was previously no `InternalServerError` method defined for `APIContext` then the one used for `Context` was used. This PR stops that by providing a method specific for these to `APIContext`.

**Review by lafriks — APPROVED** [review]

**Comment by lafriks** [comment]
sorry my bad :)

**Comment by 6543** [comment]
I would say ready to merge

**Comment by guillep2k** [comment]
Please send backport. Can you check whether the "Allow maintainers to edit this post" checkbox shows and is checked up before posting? It's an ongoing mistery.... 🕵️

**Comment by 6543** [comment]
I agree It's an ongoing mistery!

**Comment by 6543** [comment]
have created backport to v1.11 too - old untuched code allowed it easely ;)

**Pull request merged**